### ISSUE TITLE
Infinity loop due to cache merging

### ISSUE THREAD

**Issue by timothyarmes** [apollographql/apollo-client#7028]
My app has a query to fetch the currently logged in user, and a HOC is used to fetch that user and render sub components.

```
const GET_SELF = gql`
  query me {
    me {
      ...userFragment
      prefs { ...userPrefsFragment }
    }
  }
  
  ${Users.userFragment}
  ${Users.userPrefsFragment}
`;

const UserProvider = ({ render }) => {
  const { error, loading, data: { me } = {}, refetch, subscribeToMore } = useQuery(GET_SELF, { fetchPolicy: 'cache-first' });

  <snip>

  if (loading) return <Loading />;
  return render(me);
}
```

Now, in a subcomponent I'm running a query which fetches a set of users:

```
export const GET_MISSION = gql`
  query mission($id: ID!) {
    mission(_id: $id) {
      users { ...userDisplayFragment }
    }
  }

 ${userDisplayFragment}
`;
```

Sometime that query will return the currently logged in user in the list, along with the requested fields.

**Intended outcome:**
The subcomponent renders.  This worked perfectly under Apollo 2.

**Actual outcome:**
The fields returned for the currently logged in user are merged into the user object in the cache. Despite the fact that these fields *have not changed* the value of the user object, this causes the 'me' query to re-render the entire tree, which reruns the 'mission' query above, which leads to an infinite loop.

I've played around endlessly with the Type Policies but I can't find a solution.

**Versions**

  System:
    OS: macOS 10.15
  Binaries:
    Node: 12.13.1 - ~/.nvm/versions/node/v12.13.1/bin/node
    Yarn: 1.19.2 - /usr/local/bin/yarn
    npm: 6.12.1 - ~/.nvm/versions/node/v12.13.1/bin/npm
  Browsers:
    Chrome: 85.0.4183.102
    Firefox: 77.0.1
    Safari: 13.0.2
  npmPackages:
    @apollo/client: ^3.2.0 => 3.2.0 
    apollo-link-logger: ^1.2.3 => 1.2.3 
    apollo-server-express: ^2.17.0 => 2.17.0

**Comment by benjamn** [comment]
@timothyarmes Can you put this setup into a [reproduction](https://github.com/apollographql/react-apollo-error-template)? That will help us give you the right advice without guessing (though I can keep guessing if a repro is not feasible).

As an initial note, that `DatePolicy` pattern seems risky because you'll be using the same `DatePolicy` instance (with only one `this.prevDate`) for all `User.createdAt` fields, even for different users. I don't think that will solve the problem, but it seemed worth mentioning.

**Comment by timothyarmes** [comment]
Hi Ben. I'll try to find time to create a repo. For the DatePolicy, I'm doing new DatePolicy() so it should be okay...

**Comment by ziggy6792** [comment]
Hey @timothyarmes

Thanks for sharing this I was having the exact same issue.
I came across [this](https://github.com/apollographql/apollo-client/issues/585#issuecomment-400762881) issue and @theodorDiaconu has done a brilliant job creating a small package that seems to do what we want here. I following his example and it works beautifully.

Good luck :)

**Comment by brainkim** [comment]
Gentle reminder that we’re looking for reproductions. I greatly prioritize any issue which I can run, and I got a lot of issues on my plate at the moment. If not, I will get to this soon no worries 😇

**Comment by rcbevans** [comment]
I'm hitting the same issue, just spent 3 hours trying to debug.

The trigger seems to be a mutation using an update function to merge a field onto a type with a read function specified on the type policy.

**Comment by garrett-gottlieb** [comment]
This took a lot of trial/error, but for me it was printing out the `options` argument for the `merge` function:
```ts
              merge(existing = [], incoming, options) {
                const merged = _.uniqBy([...existing, ...incoming], '__ref')

                console.log('Merged audioPosts', {
                  existing,
                  incoming,
                  options,
                  variables, // ADDING `: options.variables,` FIXED THE ISSUE (along with a React Native printing error)
                  merged,
                })

**Comment by wallisonsilvarepassa** [comment]
Your UserType or fields within UserType (createdAt) probably don't have an ID for Apollo to deduplicate the data in the cache.
https://www.apollographql.com/docs/kotlin/caching/troubleshooting

What is probably happening is that Apollo saves the data of the logged User in the cache and the other request that lists the users modifies the data of this logged user (with different data) forcing a cache update. A single missing field in the cache for the query querying the logged in user will cause a new request. The infinite loop happens because the UserType data in different queries are different and each request updates the cache with the expected format.

**Comment by WallisonHenrique** [comment]
Your UserType or fields within UserType (createdAt) probably don't have an ID for Apollo to deduplicate the data in the cache.
https://www.apollographql.com/docs/kotlin/caching/troubleshooting

What is probably happening is that Apollo saves the data of the logged User in the cache and the other request that lists the users modifies the data of this logged user (with different data) forcing a cache update. A single missing field in the cache for the query querying the logged in user will cause a new request. The infinite loop happens because the UserType data in separate queries are different and each request updates the cache with the expected format.

**Comment by alessbell** [comment]
Hi everyone 👋  In case it's helpful, we have a [CodeSandbox](https://codesandbox.io/s/github/apollographql/react-apollo-error-template) and forkable [GitHub repo](https://github.com/apollographql/react-apollo-error-template) for creating reproductions, which would greatly aid in prioritizing/assessing this issue. Thanks!

**Comment by marco2216** [comment]
I experienced almost exactly the same issue today. Adding a read function to a field would cause two queries to be fetched in an infinite loop.
I figured out that it was caused by one of those two queries that queried data in this structure:
```
 product {
        id
        details {
          id
          type
          title
        }
      }
```
The backend returned the same id for `product` and `details`, and our default `dataIdFromObject` function just returns the object id directly, so they would have the same cache id. Adding the typename to the `dataIdFromObject` function fixed the issue.
Quite an odd issue, the read field was added to a completely different field, a field that was not even queried by the query that the snippet above is from. It was queried by the other query which was a part of the infinite loop though.

**Comment by marco2216** [comment]
It started happening again with different queries, now I realised that the infinite loop also happens if there are some queries getting data from types that don't have ID where there is no merge function defined in the cache, if the queries override the data of each other.
For some reason, this doesn't cause any issues usually, but as soon as the read function that returns new data is added, if one of the queries queries the field the read function is added to, it starts causing the infinite loop.

**Comment by marco2216** [comment]
@alessbell 
Here is a reproduction of the issue with the cause being that different data in two queries is being fetched from the same type that doesn't have an ID/ or merge function in the cache combined with the custom read function on `Person.name`
Upon opening the Sandbox, if you open the browser console, you should see `query AllPeople` being logged continuously, since the query is being repeated infinitely.
If you remove line 121 in `index.jsx` `name: () => "name_",` and reload the page, you should see that the query only fires once.
[Sandbox link](https://codesandbox.io/p/sandbox/dark-dew-ym3k82?file=%2Fsrc%2Findex.jsx%3A121%2C34&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloe891rc00083b6in3aioh89%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloe891rc00033b6i49k2du1p%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloe891rc00053b6i4zaox0ub%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloe891rc00073b6isz5o1grt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloe891rc00033b6i49k2du1p%2522%253A%257B%2522id%2522%253A%2522cloe891rc00033b6i49k2du1p%2522%252C%2522activeTabId%2522%253A%2522cloe8f77o03ec3b6iiie3ad73%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloe891rc00023b6imgpniczt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fschema.js%2522%252C%2522id%2522%253A%2522cloe8b8z5008p3b6iepnv30tu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Flink.js%2522%252C%2522id%2522%253A%2522cloe8et1n033n3b6ipy794bw3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.jsx%2522%252C%2522id%2522%253A%2522cloe8f77o03ec3b6iiie3ad73%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cloe891rc00073b6isz5o1grt%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloe891rc00063b6iykhviadl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cloe891rc00073b6isz5o1grt%2522%252C%2522activeTabId%2522%253A%2522cloe891rc00063b6iykhviadl%2522%257D%252C%2522cloe891rc00053b6i4zaox0ub%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloe891rc00043b6in4ta0sgi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%255D%252C%2522id%2522%253A%2522cloe891rc00053b6i4zaox0ub%2522%252C%2522activeTabId%2522%253A%2522cloe891rc00043b6in4ta0sgi%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by phryneas** [comment]
Hi @marco2216, I've taken a look at this and believe #11430 might get closer to solving this problem. The problem of potential feuds between queries is very complex, though, so this will likely need a lot more investigation before we can get it in - just as a fair warning.

**Comment by jerelmiller** [comment]
Hey all 👋 

Thanks for the patience! I've been doing a bit with cache feuds and came across this issue. I've tested the reproduction from https://github.com/apollographql/apollo-client/issues/7028#issuecomment-1787038479 in the latest versions of Apollo Client (4.x) and I don't see the infinite refetch issue so I think this might have been fixed.

That said, I just opened https://github.com/apollographql/apollo-client/pull/13406 (which will be released in the upcoming 4.3 release) which makes a couple changes to how feud detection works, especially as it relates to `read` functions. Currently the transformed values from `read` functions aren't applied to the result after that automatic refetch. I can confirm this with the reproduction above which shows `name_` for the first query, but the 2nd query shows the raw network values. https://github.com/apollographql/apollo-client/pull/13406 now makes sure the `read` function values are applied. I've also added a new warning that shows up when feud detection kicks in. Here is what that looks like in the reproduction from https://github.com/apollographql/apollo-client/issues/7028#issuecomment-1787038479.

<img width="1327" height="205" alt="Image" src="https://github.com/user-attachments/assets/a5f1aefd-19fc-4196-979a-0018a57b3203" />

Unless I hear otherwise, I'm going to consider this fixed by https://github.com/apollographql/apollo-client/pull/13406 and close this issue once that is merged. Hope this at least provides a bit better experience!

**Comment by jerelmiller** [comment]
I just published [`4.3.0-alpha.7`](https://github.com/apollographql/apollo-client/releases/tag/%40apollo%2Fclient%404.3.0-alpha.7) which contains the warning for feuds. Since I believe everything is addressed given the original issue description, I'm going to go ahead and close this as completed. If you try the alpha and see an issue, please open a new issue and I'd be happy to dig in further. Thanks!
